Make SMP CheriBSD boot in QEMU

nwf opened the pull request:
This is almost entirely Jess's work, I just push buttons when told to do so.

jrtc27 committed:
Enable MT ASE for default CHERI CPU

Used by FreeBSD's Malta SMP support

nwf committed:
MIPS 5Kf: also assert config3 MT

A follow-on to Jess' "Enable MT ASE for default CHERI CPU"

nwf committed:
MIPS BERI: enable MT

arichardson commented on target/mips/helper.c:
```diff
@@ -1762,6 +1762,19 @@ static inline ram_addr_t v2r_addr(CPUMIPSState *env, target_ulong vaddr, MMUAcce
+        /* Check RAM address to see if the linkedflag needs to be reset. */
+        if (paddr <= other_env->lladdr && other_env->lladdr < paddr + size)
+            other_env->linkedflag = 0;
+    }
```
We should really move away from using the linkedflag hack. But that can be done some other time.
The next time we merge from upstream, the ll/sc handling in the MIPS target will have changed so we should probably use the same approach for CHERI atomics.

nwf replied:
Have a convenient pointer for what upstream does now?

arichardson replied:
Hmm we might actually have merged this already. 
Anyway, linkedflag was added only for CHERI and means that CHERI cll* and MIPS ll* use separate linked statuses so I believe we can have both ll/sc and cllc/cscc succeed even if they touch the same address...

arichardson reviewed (approved):
Looks good to me if it works 👍

nwf commented:
Oh dear.  No, please don't merge this.  QEMU... attempts to emulate LL/SC using... CAS as of 33a07fa2db6.  That's **emphatically** not how that works and means that the MIPS code that doesn't guard against ABA, because it shouldn't have to, is going to break in subtle but thourough ways.  ARGH.

arichardson commented:
We explicitly disable multi-TCG, so I don't think we can run into ABA problems?

nwf commented:
Is that right?  We might switch between threads between the LL and SC if there's a jump between them that breaks up the basic block (as there would be in, for example, the LL/SC CAS loop).

It's probably best to merge it and just acknowledge that SMP emulation is not quite faithful, even so.

arichardson commented:
> Is that right? We might switch between threads between the LL and SC if there's a jump between them that breaks up the basic block (as there would be in, for example, the LL/SC CAS loop).
> 
> It's probably best to merge it and just acknowledge that SMP emulation is not quite faithful, even so.

I think you're right, it could still break unless we unset the flag on emulated CPU switch. I'm not sure if there is a callback that we can modify to unset it.

nwf commented:
4ae16a125eee8fbb2ef3cbfbea291f850508467d no longer applies after the refactoring to move capability code out to common files.  But looking at that patch in more detail, I don't think it was doing the right thing, because `->lladdr` is a virtual address (and `CP0_LLAddr` is the physical address, at least according to `cllc_without_tcg`).

But looking around the CHERI LL/SC code is a house of horrors due to the existence of `CP0_LLAddr`, `lladdr`, `linkedflag` and the use of CAS rather than actually DTRT with LL/SC.  The DDC-relative/integer-address conditional stores use `lladdr` and CAS but the cap-relative conditional stores use `linkedflag`, so it's not even that "capabilities and data" are speaking different links; *all* the CHERI instructions are funky.

Upstream's assertion of supporting MT TCG is a lie due to their CAS emulation, but I suspect fixing that is a bridge too far.  http://www.cs.columbia.edu/~cota/pubs/cota_cgo17.slides.pdf suggests two fixes (Pico-ST and Pico-HTM), both of which sound unpleasant.

At the very least we should harmonize the cap-relative instructions to use the terrible-awful-no-good CAS-based emulation of LL/SC (ick), which, I think, means increasing `llval` to a `__int128` from `target ulong`.  All the machines we routinely run on should understand 16-byte wide CAS operations where necessary.

On the other hand, given that we have to serialize for the tag table anyway, maybe just a rwlocked scoreboard for linked addresses isn't that much worse.

Thoughts?

nwf committed:
CHERI: Overhaul LL/SC atomics

CHERI QEMU, for rather fundamental reasons of its implementation, does
not support MTTCG.  Therefore, we can adopt a very simple, crude
approximation of capability-mediated LL/SC: the current CPU core holds
any link until the core is preempted or until guest software does
something the explicitly breaks it.

nwf commented:
Because we're not getting away from !MTTCG any time soon, it might make sense to just have our atomics come along for the ride.  To that end, use the end-of-TCG-code callback (`mips_cpu_synchronize_from_tb`) to nerf `lladdr` so that subsequent store-conditionals fail.

While here, make it no longer the responsibility of the capability marshalling and tag table logic to break load link state.  Finally get rid of that `linkedflag` thing, not that this is necessarily the *right* way to do that...

For whatever it's worth, as hideous as it may be, it passes `cheritest` and boots CheriBSD in UP, so things are at least not too much worse than they are now.  In 2-way SMP, even with the rest of #76 and `options SMP` turned on, the second CPU is failing to come online:

```
avail memory = 2073808896 (1977MB)
Waiting for cpu1 to start
CPU 1 failed to start
AP #1 failed to start: -2
FreeBSD/SMP: Multiprocessor System Detected: 2 CPUs
random: no preloaded entropy cache
```

arichardson commented on target/riscv/translate.c:
```diff
@@ -1173,3 +1173,10 @@ void riscv_translate_init(void)
+#ifdef TARGET_CHERI
+void gen_cheri_break_loadlink(TCGv_cap_checked_ptr out_addr, TCGv_i32 memop)
+{
+    // TODO
```
```suggestion
    tcg_gen_movi_tl(load_res, -1);
```

arichardson commented on tcg/tcg-op.c:
```diff
@@ -2821,6 +2821,7 @@ static inline void plugin_gen_mem_callbacks(TCGv_cap_checked_ptr vaddr, uint16_t
 #ifdef TARGET_CHERI
 static inline void gen_cheri_invalidate_tags(TCGv_cap_checked_ptr out_addr, TCGv_i32 memop) {
     gen_helper_cheri_invalidate_tags(cpu_env, out_addr, memop);
+    gen_cheri_break_loadlink(out_addr, memop);
```
Maybe we should do the `gen_cheri_break_loadlink` unconditionally (i.e. even for MIPS without CHERI). That would allow reverting https://github.com/CTSRD-CHERI/cheritest/commit/3d5146e450a9beb7c688763a3ec18702f3bc344d and https://github.com/CTSRD-CHERI/cheritest/commit/494c0dd2354215e3f53475d36fc1319e9000dc23

nwf replied:
OK; here's a version that tries to do just that.  This continues to pass `cheritest`.

arichardson commented on target/mips/cpu.c:
```diff
@@ -207,6 +207,20 @@ static void mips_cpu_synchronize_from_tb(CPUState *cs, TranslationBlock *tb)
     env->hflags &= ~MIPS_HFLAG_BMASK;
     env->hflags |= tb->flags & MIPS_HFLAG_BMASK;
+
+#ifdef TARGET_CHERI
```
We could probably remove the TARGET_CHERI? I think clearing it for the current CPU is also fine if you have mttcg.

arichardson reviewed (approved):
I'm surprised about the jenkins failure, I can't see the error in the logfile...
I'll try to run CHERI+MIPS64 tests as two separate steps, maybe that helps.

But other than than LGTM. Thanks!

nwf commented:
OK, I have done some further debugging as to why this isn't booting CheriBSD in qemu `-smp 2`... and something is horked in the bootloader, even before we get to any CheriBSD code, so clearly something has gone terribly awry.

CPU 0 bringup through the bootloader looks like this:
```
Trace 0: 0x7fff44000180 [0000000000000000/ffffffffbfc00000/ffffffffffffffff/0x7d/0x4000098] 
pc=0xffffffffbfc00000 HI=0x0000000000000000 LO=0x0000000000000000 ds 4000098 0000000000000000 0
GPR00: zero 0000000000000000 at 0000000000000000 v0 0000000000000000 v1 0000000000000000
GPR04: a0 0000000000000000 a1 0000000000000000 a2 0000000000000000 a3 0000000000000000
GPR08: a4 0000000000000000 a5 0000000000000000 a6 0000000000000000 a7 0000000000000000
GPR12: t0 0000000000000000 t1 0000000000000000 t2 0000000000000000 t3 0000000000000000
GPR16: s0 0000000000000000 s1 0000000000000000 s2 0000000000000000 s3 0000000000000000
GPR20: s4 0000000000000000 s5 0000000000000000 s6 0000000000000000 s7 0000000000000000
GPR24: t8 0000000000000000 t9 0000000000000000 k0 0000000000000000 k1 0000000000000000
GPR28: gp 0000000000000000 sp 0000000000000000 s8 0000000000000000 ra 0000000000000000
CP0 Status  0x00400080 Cause   0x00000000 EPC    0x0000000000000000
    Config0 0x8000c082 Config1 0xbe6130db LLAddr 0x0000000000000000
    Config2 0x80000000 Config3 0x0c002084
    Config4 0x00000000 Config5 0x00000000
[0:0] 0xffffffffbfc00000:  j    0xffffffffbfc00580
Linking TBs 0x7fff44000180 [ffffffffbfc00000] index 0 -> 0x7fff44000320 [ffffffffbfc00580]
Trace 0: 0x7fff44000320 [0000000000000000/ffffffffbfc00580/ffffffffffffffff/0x7d/0x4000098] 
pc=0xffffffffbfc00580 HI=0x0000000000000000 LO=0x0000000000000000 ds 4000098 0000000000000000 0
GPR00: zero 0000000000000000 at 0000000000000000 v0 0000000000000000 v1 0000000000000000
GPR04: a0 0000000000000000 a1 0000000000000000 a2 0000000000000000 a3 0000000000000000
GPR08: a4 0000000000000000 a5 0000000000000000 a6 0000000000000000 a7 0000000000000000
GPR12: t0 0000000000000000 t1 0000000000000000 t2 0000000000000000 t3 0000000000000000
GPR16: s0 0000000000000000 s1 0000000000000000 s2 0000000000000000 s3 0000000000000000
GPR20: s4 0000000000000000 s5 0000000000000000 s6 0000000000000000 s7 0000000000000000
GPR24: t8 0000000000000000 t9 0000000000000000 k0 0000000000000000 k1 0000000000000000
GPR28: gp 0000000000000000 sp 0000000000000000 s8 0000000000000000 ra 0000000000000000
CP0 Status  0x00400080 Cause   0x00000000 EPC    0x0000000000000000
    Config0 0x8000c082 Config1 0xbe6130db LLAddr 0x0000000000000000
    Config2 0x80000000 Config3 0x0c002084
    Config4 0x00000000 Config5 0x00000000
[0:0] 0xffffffffbfc00004:  nop
[0:0] 0xffffffffbfc00580:  li   a0,2
    Write a0 = 0000000000000002
[0:0] 0xffffffffbfc00584:  lui  sp,0x8000
    Write sp = ffffffff80000000
[0:0] 0xffffffffbfc00588:  ori  sp,sp,0x1fc0
    Write sp = ffffffff80001fc0
[0:0] 0xffffffffbfc0058c:  lui  a1,0x8000
    Write a1 = ffffffff80000000
[0:0] 0xffffffffbfc00590:  ori  a1,a1,0x2000
    Write a1 = ffffffff80002000
[0:0] 0xffffffffbfc00594:  lui  a2,0x8000
    Write a2 = ffffffff80000000
[0:0] 0xffffffffbfc00598:  ori  a2,a2,0x2008
    Write a2 = ffffffff80002008
[0:0] 0xffffffffbfc0059c:  lui  a3,0x200
    Write a3 = 0000000002000000
[0:0] 0xffffffffbfc005a0:  ori  a3,a3,0x0
    Write a3 = 0000000002000000
[0:0] 0xffffffffbfc005a4:  lui  a5,0xb400
    Write a5 = ffffffffb4000000
[0:0] 0xffffffffbfc005a8:  lui  a4,0xdf00
    Write a4 = ffffffffdf000000
[0:0] 0xffffffffbfc005ac:  sw   a4,104(a5)
    Memory Write [ffffffffb4000068] = df000000
```

and we're off to the races.  But CPU1's bringup through the same code looks like this:

```
Trace 1: 0x7fff44000180 [0000000000000000/ffffffffbfc00000/ffffffffffffffff/0x7d/0x4000098] 
pc=0xffffffffbfc00000 HI=0x0000000000000000 LO=0x0000000000000000 ds 4000098 0000000000000000 0
GPR00: zero 0000000000000000 at 0000000000000000 v0 0000000000000000 v1 0000000000000000
GPR04: a0 0000000000000000 a1 0000000000000000 a2 0000000000000000 a3 0000000000000000
GPR08: a4 0000000000000000 a5 0000000000000000 a6 0000000000000000 a7 0000000000000000
GPR12: t0 0000000000000000 t1 0000000000000000 t2 0000000000000000 t3 0000000000000000
GPR16: s0 0000000000000000 s1 0000000000000000 s2 0000000000000000 s3 0000000000000000
GPR20: s4 0000000000000000 s5 0000000000000000 s6 0000000000000000 s7 0000000000000000
GPR24: t8 0000000000000000 t9 0000000000000000 k0 0000000000000000 k1 0000000000000000
GPR28: gp 0000000000000000 sp 0000000000000000 s8 0000000000000000 ra 0000000000000000
CP0 Status  0x00400080 Cause   0x40008000 EPC    0x0000000000000000
    Config0 0x8000c082 Config1 0xbe6130db LLAddr 0x0000000000000000
    Config2 0x80000000 Config3 0x0c002084
    Config4 0x00000000 Config5 0x00000000
[1:0] 0xffffffffbfc00000:  j    0xffffffffbfc00580
[1:0] 0xffffffffbfc00004:  nop
[1:0] 0xffffffffbfc00580:  li   a0,2
    Write a0 = 0000000000000002
[1:0] 0xffffffffbfc00584:  lui  sp,0x8000
    Write sp = ffffffff80000000
[1:0] 0xffffffffbfc00588:  ori  sp,sp,0x1fc0
    Write sp = ffffffff80001fc0
[1:0] 0xffffffffbfc0058c:  lui  a1,0x8000
    Write a1 = ffffffff80000000
[1:0] 0xffffffffbfc00590:  ori  a1,a1,0x2000
    Write a1 = ffffffff80002000
[1:0] 0xffffffffbfc00594:  lui  a2,0x8000
    Write a2 = ffffffff80000000
[1:0] 0xffffffffbfc00598:  ori  a2,a2,0x2008
    Write a2 = ffffffff80002008
[1:0] 0xffffffffbfc0059c:  lui  a3,0x200
    Write a3 = 0000000002000000
[1:0] 0xffffffffbfc005a0:  ori  a3,a3,0x0
    Write a3 = 0000000002000000
[1:0] 0xffffffffbfc005a4:  lui  a5,0xb400
    Write a5 = ffffffffb4000000
[1:0] 0xffffffffbfc005a8:  lui  a4,0xdf00
    Write a4 = ffffffffdf000000
Trace 1: 0x7fff44658060 [0000000000000000/ffffffffbfc00380/ffffffffffffffff/0x7d/0x4000098] 
pc=0xffffffffbfc00380 HI=0x0000000000000000 LO=0x0000000000000000 ds 4000098 0000000000000000 0
GPR00: zero 0000000000000000 at 0000000000000000 v0 0000000000000000 v1 0000000000000000
GPR04: a0 0000000000000002 a1 ffffffff80002000 a2 ffffffff80002008 a3 0000000002000000
GPR08: a4 ffffffffdf000000 a5 ffffffffb4000000 a6 0000000000000000 a7 0000000000000000
GPR12: t0 0000000000000000 t1 0000000000000000 t2 0000000000000000 t3 0000000000000000
GPR16: s0 0000000000000000 s1 0000000000000000 s2 0000000000000000 s3 0000000000000000
GPR20: s4 0000000000000000 s5 0000000000000000 s6 0000000000000000 s7 0000000000000000
GPR24: t8 0000000000000000 t9 0000000000000000 k0 0000000000000000 k1 0000000000000000
GPR28: gp 0000000000000000 sp ffffffff80001fc0 s8 0000000000000000 ra 0000000000000000
CP0 Status  0x00400082 Cause   0x4000801c EPC    0xffffffffbfc005ac
    Config0 0x8000c082 Config1 0xbe6130db LLAddr 0x0000000000000000
    Config2 0x80000000 Config3 0x0c002084
    Config4 0x00000000 Config5 0x00000000
[1:0] 0xffffffffbfc005ac:  sw   a4,104(a5)
-> Switch to Kernel mode
-> Exception #7 vector 0xffffffffbfc00380 fault-addr 0x0000000000000000
    Write PCC|v:1 s:0 p:00078fff f:0 b:0000000000000000 l:ffffffffffffffff
             |o:ffffffffbfc00380 t:3ffff
    Write EPCC|v:1 s:0 p:00078fff f:0 b:0000000000000000 l:ffffffffffffffff
             |o:ffffffffbfc005ac t:3ffff
    Write ErrorEPCC|v:1 s:0 p:00078fff f:0 b:0000000000000000 l:ffffffffffffffff
             |o:ffffffffbfc00000 t:3ffff
    Write ErrorEPC = ffffffffbfc00000
do_raise_exception_err: data bus error 0
mips_cpu_do_interrupt enter: PC ffffffffbfc005ac EPC 0000000000000000 data bus error exception, (hflags & MIPS_HFLAG_BMASK)=0, hflags=4000098
        PCC=v:1 s:0 p:00078fff f:0 b:0000000000000000 l:ffffffffffffffff o:ffffffffbfc005ac t:3ffff
        KCC= v:1 s:0 p:00078fff f:0 b:0000000000000000 l:ffffffffffffffff o:0000000000000000 t:3ffff
        EPCC=v:1 s:0 p:00078fff f:0 b:0000000000000000 l:ffffffffffffffff o:0000000000000000 t:3ffff
mips_cpu_do_interrupt: PC ffffffffbfc00380 EPC ffffffffbfc005ac cause 7
    S 00400082 C 4000801c A 0000000000000000 D 0000000000000000
```

And the CPU ends up stuck in a loop faulting at `0xffffffffbfc005ac`. :(

nwf commented:
OK, with CTSRD-CHERI/cheribsd#541 in tree, this looks like it works and should be good to merge.

arichardson commented on target/riscv/translate.c:
```diff
@@ -1173,3 +1173,8 @@ void riscv_translate_init(void)
+
+void gen_cheri_break_loadlink(TCGv_cap_checked_ptr out_addr, TCGv_i32 memop)
+{
+    tcg_gen_movi_tl((TCGv)load_res, -1);
```
Seems like this line is causing a fatal error during RISC-V QEMU boot.
```
qemu-system-riscv64cheri: FATAL: cheri_tag_invalidate: ffffffffffffffff+8 crosses a page boundary
qemu-system-riscv64cheri: Probably caused by guest instruction: 0xffffffc000278736:  18a4b62f          sc.d            a2,a0,(s1)
```

It seems like the RISC-V SC implementation uses load_res directly (rather than the argument to sc) to avoid needing a local TCGTemp (since it needs to survive past the addr==load_res check):
```
    tcg_gen_atomic_cmpxchg_tl_with_checked_addr(src1, load_res, load_val, src2, ctx->mem_idx, mop);
```

RISC-V regression tests running for each pull request would be nice 😢 .
I'll comment this out for now and will look into a fix tomorrow.
